drivers/flash/spi_nand.c

```diff
@@ -0,0 +1,1026 @@
+		size -= chunk;
+	}
+
+out:	release_device(dev);
+
+	return ret;
+}
+
+static int spi_nand_read_software_ecc(const struct device *dev, 
+			off_t addr, void *dest, size_t size)
+{
+	struct spi_nand_data *data = dev->data;
+	int ret = 0;
+	uint32_t offset = 0;
+	uint32_t chunk = 0;
+	uint32_t read_bytes = 0;
+	uint8_t ecc_steps = data->ecc_steps;
+	uint8_t *p = (uint8_t *)data->page_buf;
+
+	acquire_device(dev);
+
+	while (size > 0) {
+		/* Read on _page_size_bytes boundaries (Default 2048 bytes a page) */
+		offset = addr % data->page_size;
+		chunk = (offset + size < data->page_size) ? 
+			size : (data->page_size - offset);
+		read_bytes = chunk;
+
+		ret = spi_nand_access(dev, SPI_NAND_CMD_PAGE_READ,
+			NAND_ACCESS_ADDRESSED | NAND_ACCESS_24BIT_ADDR,
+			addr >> data->page_shift, NULL, 0);
+		if (ret != 0) {
+			LOG_ERR("page read failed: %d", ret);
+			goto out;
+		}
+
+		ret = spi_nand_wait_until_ready(dev);
+
+		ret = spi_nand_access(dev, SPI_NAND_CMD_READ_CACHE,
+			NAND_ACCESS_ADDRESSED | NAND_ACCESS_16BIT_ADDR | NAND_ACCESS_DUMMY,
+			(addr & SPI_NAND_PAGE_MASK), data->page_buf,
+			data->page_size + data->oob_size);
+		if (ret != 0) {
+			LOG_ERR("read from cache failed: %d", ret);
+			goto out;
+		}
+
+		memcpy(data->ecc_code, data->page_buf + data->page_size + data->ecc_layout_pos,
+			data->ecc_bytes * data->ecc_steps);
+
+		p = (uint8_t *)data->page_buf;
+		ecc_steps = data->ecc_steps;
+		for (uint8_t i = 0 ; ecc_steps; ecc_steps--, i += data->ecc_bytes, p += data->ecc_size) {
+			memset(data->nbc.bch->input_data, 0x0, (1 << data->nbc.bch->m) / 8);
+			memcpy(data->nbc.bch->input_data + data->ecc_bytes, p, data->ecc_size);
+
+			int ret = bch_decode(data->nbc.bch, data->nbc.bch->input_data, (unsigned int *)(data->ecc_code + i));
+			if (ret < 0) {
+				LOG_ERR("Reading data failed");
+				return -ENODEV;
+			}
+			memcpy(p, data->nbc.bch->input_data + data->ecc_bytes, data->ecc_size);
+		}
+
+		memcpy(dest, data->page_buf + offset, read_bytes);
+
+		dest = (uint8_t *)dest + chunk;
+
+		addr = (addr + SPI_NAND_PAGE_OFFSET) & (~SPI_NAND_PAGE_MASK);
+		size -= chunk;
+	}
+
+out:	release_device(dev);
+
+	return ret;
+}
+
+static int spi_nand_read(const struct device *dev, off_t addr, void *dest,
+			size_t size)
+{
+	int ret = 0;
+	struct spi_nand_data *data = dev->data;
+
+	if (data->ecc_bits == 0) {
+		if (data->continuous_read) {
+			ret = spi_nand_read_cont(dev, addr, dest, size);
+		} else {
+			ret = spi_nand_read_normal(dev, addr, dest, size);
+		}
+	} else {
+		ret = spi_nand_read_software_ecc(dev, addr, dest, size);
+	}
+
+	return ret;
+}
+
+static int spi_nand_write(const struct device *dev, off_t addr,
+			 const void *src,
+			 size_t size)
+{
+	struct spi_nand_data *data = dev->data;
+	int ret = 0;
+	uint32_t offset = 0;
+	uint32_t chunk = 0;
+	uint32_t written_bytes = 0;
+
+	/* address must be sub-page-aligned */
+	if (!(addr & (SPI_NAND_SUB_PAGE_SIZE - 1)) && (addr != 0)) {
```

Comment:
if (!(addr & (SPI_NAND_SUB_PAGE_SIZE - 1)) && (addr != 0)) 

contains incorrect logic. You should remove the first logical not ("!") to fix it. 

correction-> if ((addr & (SPI_NAND_SUB_PAGE_SIZE - 1)) && (addr != 0)) 

otherwise it doesn't accept page aligned writes. 

address 262144 (page 0 of block 1) returns a true and resolves an error which shouldn't be the case.

Reply:
could just be:
if (addr & SPI_NAND_BLOCK_MASK)

Reply:
resolved